birdhouse/scripts/read-configs.include.sh

```diff
@@ -166,19 +177,76 @@ read_components_default_env() {
 
     # EXTRA_CONF_DIRS and DEFAULT_CONF_DIRS relative paths are relative to COMPOSE_DIR.
     if [ -d "$COMPOSE_DIR" ]; then
-        cd "$COMPOSE_DIR"
+        cd "$COMPOSE_DIR" >/dev/null
     fi
 
     source_conf_files "$DEFAULT_CONF_DIRS" 'DEFAULT_CONF_DIRS'
     source_conf_files "$EXTRA_CONF_DIRS" 'EXTRA_CONF_DIRS'
 
     # Return to previous pwd.
     if [ -d "$COMPOSE_DIR" ]; then
-        cd -
+        cd - >/dev/null
     fi
 }
 
 
+# Check that all optional variables are defined with a different value than the default to emit a warning log message.
+# Also check that required variables do not use generic defaults to indicate possible security issues.
+check_default_vars() {
+    # for required variables, do not check for omitted override
+    # those will be flagged as error (check_required_vars)
+    # only indicate if there is a possible security concern
+    # note that the defaults of required variables are not define in those variables
+    # (ie: __DEFAULT__MAGPIE_ADMIN_PASSWORD exists, but MAGPIE_ADMIN_PASSWORD is not set, must have explicit override)
+    for i in ${VARS}
+    do
+      n="${i#\$}"
+      v=`eval echo "${i}" 2>/dev/null`
+      default="\${__DEFAULT__${n}}"
+      d=`eval echo "${default}" 2>/dev/null`
+      if [ ! -z "${d}" ]; then
+          if [ "${d}" = "${v}" ]; then
+              log WARN \
+                "Required variable [${n}] employs a default recommended for override." \
+                "The security of your deployment may be compromised unless it is changed. Check env.local file."
+          fi
+      fi
+    done
+
+    # for optional variables, warn about possibility omitted override or when defaults are employed
+    for i in ${OPTIONAL_VARS}
+    do
+        v="${i}"
+        d=`eval echo "$v"`
+        n="${i#\$}"
+        default="\${__DEFAULT__${n}}"
+        result=`echo "${d}" | grep -c "${default}"`
+        if [ -z "`eval "echo ${v}"`" ]
+        then
+            log WARN "Optional variable [${n}] is not set. Check env.local file."
+        fi
+        if [ "${result}" -gt 0 ]
+        then
+            log WARN "Optional variable [${n}] employs a default recommended for override. Check env.local file."
+        fi
+    done
+}
+
+
+# Verify that all required variables are set, and error out otherwise with an error log message.
+check_required_vars() {
+    for i in ${VARS}
+    do
+        v="${i}"
+        if [ -z "`eval "echo ${v}"`" ]
+        then
+            log ERROR "Required variable $v is not set. Check env.local file."
+            exit 1
```

Comment:
I like that we've factored out this function into this file. It keeps all of the logic to do with reading/verifying the config variables in one file.

> Having an exit in a function is like having a sys.exit in a python module, it's something we usually avoid to do.

I don't think that we should move it to `pavics-compose.sh` but we can have it `return 1` instead of `exit 1` and then call it as:

```sh
check_required_vars || exit $?
```

Reply:
Indeed, a `return 1` works too.  As long as no `exit` in helper functions.

Reply:
I moved it because I found it used very similar structure to other variable checking utilities, although only used by `pavics-compose` for now.
Refactor to `return 1` works for me.